# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLErrorQueue.h
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 dom/canvas/WebGLErrorQueue.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/canvas/WebGLErrorQueue.h	Wed Jan 08 22:19:14 2020 +0000
@@ -0,0 +1,58 @@
+/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef WEBGLERRORQUEUE_H_
+#define WEBGLERRORQUEUE_H_
+
+#include "WebGLCommandQueue.h"
+
+namespace mozilla {
+
+class ClientWebGLContext;
+
+using mozilla::webgl::PcqStatus;
+
+enum WebGLErrorCommand {
+  CreationError,
+  OnLostContext,
+  OnRestoredContext,
+  Warning
+};
+
+/**
+ * This is the source for the WebGL error and warning queue.  When an
+ * asynchronous error or warning occurs in the host, such as the GL context
+ * being lost or a diagnostic message being printed, it is inserted into this
+ * queue to be asynchronously retrieved by the client in the content process,
+ * where it is communicated to the relevant DOM objects.
+ */
+using HostWebGLErrorSource = CommandSource<WebGLErrorCommand>;
+
+/**
+ * This is the sink for the WebGL error and warning queue.
+ * Before this sink can handle events, it needs to have its ClientWebGLContext
+ * set.
+ */
+class ClientWebGLErrorSink : public CommandSink<WebGLErrorCommand> {
+ public:
+  ClientWebGLErrorSink(UniquePtr<Consumer>&& aConsumer);
+
+  void SetClientWebGLContext(RefPtr<ClientWebGLContext>& aClientContext);
+
+ protected:
+  bool DispatchCommand(WebGLErrorCommand command) override;
+
+  RefPtr<ClientWebGLContext> mClientContext;
+};
+
+/**
+ * Factory for the WebGL error and warning queue.
+ */
+using WebGLErrorQueue =
+    CommandQueue<HostWebGLErrorSource, ClientWebGLErrorSink>;
+
+}  // namespace mozilla
+
+#endif  // WEBGLERRORQUEUE_H_